jmclient/jmclient/wallet_service.py

```diff
@@ -389,22 +399,26 @@ def transaction_monitor(self):
             if len(added_utxos) > 0 or len(removed_utxos) > 0 \
                     or txid in self.active_txs:
                 if confs == 0:
-                    for k in possible_keys:
-                        if k in self.callbacks["unconfirmed"]:
-                            for f in self.callbacks["unconfirmed"][k]:
-                                # True implies success, implies removal:
-                                if f(txd, txid):
-                                    self.callbacks["unconfirmed"][k].remove(f)
-                                    # keep monitoring for conf > 0:
-                                    self.active_txs[txid] = txd
+                    callbacks = [f for f in
+                            self.callbacks["unconfirmed"].pop(txid, [])
+                            if not f(txd, txid)]
+                    if callbacks:
+                        self.callbacks["unconfirmed"][txid] = callbacks
+                    else:
+                        self.callbacks["unconfirmed"].pop(txos, None)
+                    if self.callbacks["confirmed"].get(txid):
+                        # keep monitoring for conf > 0:
+                        self.active_txs[txid] = txd
                 elif confs > 0:
-                    for k in possible_keys:
-                        if k in self.callbacks["confirmed"]:
-                            for f in self.callbacks["confirmed"][k]:
-                                if f(txd, txid, confs):
-                                    self.callbacks["confirmed"][k].remove(f)
-                                    if txid in self.active_txids:
-                                        self.active_txs.pop(txid, None)
```

Comment:
@AdamISZ: Two bugs here as well:
* Same bug as above, w.r.t. removing an element from a list while iterating over that list.
* The transaction is removed from `self.active_txs` (formerly, `self.active_txids`) if any of the `confirmed` callbacks returned `True`, even if some `confirmed` callbacks returned `False`. In such case, the remaining callbacks will never be invoked again.

Reply:
OK, I see, very similar logic errors to the previous. Thanks.